oracle/price-feeder/oracle/oracle.go

```diff
@@ -0,0 +1,581 @@
+	for k, v := range o.prices {
+		chainDenom := o.chainDenomMapping[k]
+		// Fills in the prices with each value in the oracle
+		prices = prices.Add(sdk.NewDecCoinFromDec(chainDenom, v))
+	}
+
+	return prices
+}
+
+// SetPrices retrieves all the prices and candles from our set of providers as
+// determined in the config. If candles are available, uses TVWAP in order
+// to determine prices. If candles are not available, uses the most recent prices
+// with VWAP. Warns the the user of any missing prices, and filters out any faulty
+// providers which do not report prices or candles within 2𝜎 of the others.
+func (o *Oracle) SetPrices(ctx context.Context) error {
+	g := new(errgroup.Group)
+	mtx := new(sync.Mutex)
+	providerPrices := make(provider.AggregatedProviderPrices)
+	providerCandles := make(provider.AggregatedProviderCandles)
+	requiredRates := make(map[string]struct{})
+
+	for providerName, currencyPairs := range o.providerPairs {
+		providerName := providerName
+		currencyPairs := currencyPairs
+
+		priceProvider, err := o.getOrSetProvider(ctx, providerName)
+		if err != nil {
+			return err
+		}
+
+		for _, pair := range currencyPairs {
+			if _, ok := requiredRates[pair.Base]; !ok {
+				requiredRates[pair.Base] = struct{}{}
+			}
+		}
+
+		g.Go(func() error {
+			prices := make(map[string]provider.TickerPrice, 0)
+			candles := make(map[string][]provider.CandlePrice, 0)
+			ch := make(chan struct{})
+			errCh := make(chan error, 1)
+
+			go func() {
+				defer close(ch)
+				prices, err = priceProvider.GetTickerPrices(currencyPairs...)
+				if err != nil {
+					telemetry.IncrCounter(1, "failure", "provider", "type", "ticker")
+					errCh <- err
+				}
+
+				candles, err = priceProvider.GetCandlePrices(currencyPairs...)
+				if err != nil {
+					telemetry.IncrCounter(1, "failure", "provider", "type", "candle")
+					errCh <- err
+				}
+			}()
+
+			select {
+			case <-ch:
+				break
+			case err := <-errCh:
+				return err
+			case <-time.After(o.providerTimeout):
+				telemetry.IncrCounter(1, "failure", "provider", "type", "timeout")
+				return fmt.Errorf("provider timed out: %s", providerName)
+			}
+
+			// flatten and collect prices based on the base currency per provider
+			//
+			// e.g.: {ProviderKraken: {"ATOM": <price, volume>, ...}}
+			mtx.Lock()
+			for _, pair := range currencyPairs {
+				success := SetProviderTickerPricesAndCandles(providerName, providerPrices, providerCandles, prices, candles, pair)
+				if !success {
+					mtx.Unlock()
+					return fmt.Errorf("failed to find any exchange rates in provider responses")
+				}
+			}
+
+			mtx.Unlock()
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		o.logger.Debug().Err(err).Msg("failed to get ticker prices from provider")
+	}
+
+	computedPrices, err := GetComputedPrices(
+		o.logger,
+		providerCandles,
+		providerPrices,
+		o.providerPairs,
+		o.deviations,
+	)
+	if err != nil {
+		return err
+	}
+
+	if len(computedPrices) != len(requiredRates) {
```

Comment:
this logic seems like the price feeder is all-or-nothing. do we want to allow validators to submit prices for whatever they can get?

Reply:
we could if we want, but I'd prefer to first add this as a TODO and get the sidecar merged in. We can already tune the config toml to exclude certain denominations and thats no issue even if we dont support all whitelisted assets, whereas this relates to the presence of rates for the denoms we have in the config. If it sounds good I'll add in a TODO for more graceful exclusion of prices that are missing

Reply:
added todo